### PR TITLE
Pre-DNAT policy support

### PR DESCRIPTION
**Pull request by neiljerram**
This is the Felix part of https://github.com/projectcalico/calico/issues/823

## Description
New feature; see https://github.com/projectcalico/calico/issues/823 for details.

## Todos
- [x] Unit tests (full coverage)

### CODE REVIEW

**Comment by neiljerram** [comment]
@fasaxc - I still have lots of testing to do here, but I think it's at a good point for you to review the product code, in case any major issues emerge from that.

File: calc/policy_sorter.go [hunk 1]
```diff
@@ -135,8 +136,12 @@ func (t tierInfo) String() string {
 	policies := make([]string, len(t.OrderedPolicies))
 	for ii, pol := range t.OrderedPolicies {
 		untracked := "t"
```
Comment (fasaxc) [inline calc/policy_sorter.go]: Rename `untracked` to `polType`, for example?

File: intdataplane/endpoint_mgr.go [hunk 2]
```diff
@@ -665,12 +693,13 @@ func (m *endpointManager) resolveHostEndpoints() {
 	m.activeIfaceNameToHostEpID = newIfaceNameToHostEpID
 	m.activeHostEpIDToIfaceNames = newHostEpIDToIfaceNames
 	m.activeHostIfaceToFiltChains = newHostIfaceFiltChains
+	m.activeHostIfaceToMangleChains = newHostIfaceMangleChains
 	m.activeHostIfaceToRawChains = newHostIfaceRawChains
 
 	// Rewrite the filter dispatch chains if they've changed.
```
Comment (fasaxc) [inline intdataplane/endpoint_mgr.go]: Out of date comment.

File: intdataplane/endpoint_mgr.go [hunk 3]
```diff
@@ -88,11 +89,12 @@ type endpointManager struct {
...
-	activeHostIfaceToRawChains  map[string][]*iptables.Chain
-	activeHostIfaceToFiltChains map[string][]*iptables.Chain
+	activeHostIfaceToRawChains    map[string][]*iptables.Chain
+	activeHostIfaceToFiltChains   map[string][]*iptables.Chain
+	activeHostIfaceToMangleChains map[string][]*iptables.Chain
 	// Dispatch chains that we've programmed for host endpoints.
-	activeHostRawDispatchChains  map[string]*iptables.Chain
-	activeHostFiltDispatchChains map[string]*iptables.Chain
+	activeHostRawDispatchChains        map[string]*iptables.Chain
+	activeHostFiltMangleDispatchChains map[string]*iptables.Chain
```
Comment (fasaxc) [inline intdataplane/endpoint_mgr.go]: Didn't spot that these chains were used in two places at first.  Worth a comment to spell that out?

File: proto/felixbackend.proto [hunk 4]
```diff
@@ -265,6 +265,7 @@ message HostEndpoint {
   repeated TierInfo untracked_tiers = 6;
   repeated string expected_ipv4_addrs = 4;
   repeated string expected_ipv6_addrs = 5;
+  repeated TierInfo pre_dnat_tiers = 7;
```
Comment (fasaxc) [inline proto/felixbackend.proto]: Can move this up to be next to the others

File: rules/endpoints.go [hunk 5]
```diff
@@ -90,11 +90,34 @@ func (r *DefaultRuleRenderer) HostEndpointToRawChains(
...
 }
 
+func (r *DefaultRuleRenderer) HostEndpointToMangleChains(
+	ifaceName string,
+	preDNATPolicyNames []string,
+) []*Chain {
+	log.WithField("ifaceName", ifaceName).Debug("Rendering pre-DNAT host endpoint chain.")
+	return r.endpointToIptablesChains(
+		preDNATPolicyNames,
+		nil, // We don't render profiles into the raw chain.
```
Comment (fasaxc) [inline rules/endpoints.go]: Out of date comment.

File: rules/static.go [hunk 6]
```diff
@@ -455,6 +452,53 @@ func (r *DefaultRuleRenderer) StaticNATOutputChains(ipVersion uint8) []*Chain {
...
+	return
+}
+
+func (r *DefaultRuleRenderer) StaticManglePreroutingChain(ipVersion uint8) *Chain {
+	rules := []Rule{}
+
+	// Accept immediately if we've already accepted this packet in the raw table.
+	rules = append(rules, r.acceptAlreadyAccepted()...)
+
+	// Set a mark on the packet if it's from a workload interface.
```
Comment (fasaxc) [inline rules/static.go]: Should we just return early if from a workload?

Comment (neiljerram) [inline rules/static.go]: Yes, good spot, that makes sense.

File: rules/static_test.go [hunk 7]
```diff
@@ -502,10 +501,6 @@ var _ = Describe("Static", func() {
 				Expect(findChain(rr.StaticFilterTableChains(ipVersion), "cali-FORWARD")).To(Equal(&Chain{
```
Comment (fasaxc) [inline rules/static_test.go]: Are we missing tests of the new mangle-table equivalents to these?

File: intdataplane/endpoint_mgr.go [hunk 8]
```diff
@@ -665,40 +706,52 @@ func (m *endpointManager) resolveHostEndpoints() {
...
-// given iptables.Table and records the updates in the activeChains map.
+// given iptables.Table(s) and records the updates in the activeChains map.
 //
-// Calculating the minimum update prevents log spam and reduces the work needed in the Table.
+// Calculating the minimum update prevents log spam and reduces the work needed in the tables.
 func (m *endpointManager) updateDispatchChains(
 	activeChains map[string]*iptables.Chain,
 	newChains []*iptables.Chain,
-	table iptablesTable,
+	tables ...iptablesTable,
```
Comment (neiljerram) [inline intdataplane/endpoint_mgr.go]: Good catch, I meant to back that out, so will do.

File: rules/endpoints.go [hunk 9]
```diff
@@ -51,137 +62,149 @@ func (r *DefaultRuleRenderer) HostEndpointToFilterChains(
...
 
 const (
-	chainTypeTracked endpointChainType = iota
+	chainTypeNormal endpointChainType = iota
 	chainTypeUntracked
+	chainTypePreDNAT
 )
 
-func (r *DefaultRuleRenderer) endpointToIptablesChains(
+func (r *DefaultRuleRenderer) endpointIptablesChain(
```
Comment (neiljerram) [inline rules/endpoints.go]: Thanks.

File: rules/static_test.go (hunk 7 above)
Comment (neiljerram) [inline rules/static_test.go]: We are indeed.  If it's OK with you, I'll note that as a todo for testing, and still merge this now?

**Comment by caseydavenport** [comment]
retest this please

**Comment by neiljerram** [comment]
@caseydavenport I will look at these calico-test failures over the weekend if I can - but unfortunately that is unlikely as I'll be in Paris; otherwise will be my priority for Monday morning.

**Comment by neiljerram** [comment]
retest this please

**Pull request merged**